--- src/clang/lib/Driver/ToolChains/Darwin.cpp.orig
+++ src/clang/lib/Driver/ToolChains/Darwin.cpp
@@ -1648,11 +1648,12 @@ void DarwinClang::AddLinkRuntimeLibArgs(const ArgList &Args,
   if (isTargetIOSBased()) {
     // If we are compiling as iOS / simulator, don't attempt to link libgcc_s.1,
     // it never went into the SDK.
-    // Linking against libgcc_s.1 isn't needed for iOS 5.0+
+    // Linking against libgcc_s.1 isn't needed for iOS 5.0+ or macOS 10.6+
     if (isIPhoneOSVersionLT(5, 0) && !isTargetIOSSimulator() &&
         getTriple().getArch() != llvm::Triple::aarch64)
       CmdArgs.push_back("-lgcc_s.1");
-  }
+  } else if (isTargetMacOSBased() && isMacosxVersionLT(10, 6))
+    CmdArgs.push_back("-lgcc_s.1");
   AddLinkRuntimeLib(Args, CmdArgs, "builtins");
 }
 
