--- src/clang/lib/Driver/ToolChains/Darwin.h.orig
+++ src/clang/lib/Driver/ToolChains/Darwin.h
@@ -694,7 +694,7 @@
   void AddLinkSanitizerLibArgs(const llvm::opt::ArgList &Args,
                                llvm::opt::ArgStringList &CmdArgs,
                                StringRef Sanitizer,
-                               bool shared = true) const;
+                               bool shared = false) const;
 
   void
   AddGnuCPlusPlusIncludePaths(const llvm::opt::ArgList &DriverArgs,
