diff --git a/compiler/rustc_codegen_ssa/src/back/link.rs b/compiler/rustc_codegen_ssa/src/back/link.rs
index d35c3b6bb18..7c665344a15 100644
--- a/compiler/rustc_codegen_ssa/src/back/link.rs
+++ b/compiler/rustc_codegen_ssa/src/back/link.rs
@@ -3075,6 +3075,8 @@ fn add_apple_link_args(cmd: &mut dyn Linker, sess: &Session, flavor: LinkerFlavo
     // Intentionally verbose to ensure that the list always matches correctly
     // with the list in the source above.
     let ld64_arch = match llvm_arch {
+        "armv6" => "armv6",
+        "armv7" => "armv7",
         "armv7k" => "armv7k",
         "armv7s" => "armv7s",
         "arm64" => "arm64",
diff --git a/compiler/rustc_target/src/spec/base/apple/mod.rs b/compiler/rustc_target/src/spec/base/apple/mod.rs
index 542c5e27849..e0ed52f402f 100644
--- a/compiler/rustc_target/src/spec/base/apple/mod.rs
+++ b/compiler/rustc_target/src/spec/base/apple/mod.rs
@@ -14,6 +14,8 @@
 #[allow(non_camel_case_types)]
 #[derive(Copy, Clone, PartialEq)]
 pub(crate) enum Arch {
+    Armv6,
+    Armv7,
     Armv7k,
     Armv7s,
     Arm64,
@@ -28,6 +30,8 @@ pub(crate) enum Arch {
 impl Arch {
     fn target_name(self) -> &'static str {
         match self {
+            Self::Armv6 => "armv6",
+            Self::Armv7 => "armv7",
             Self::Armv7k => "armv7k",
             Self::Armv7s => "armv7s",
             Self::Arm64 => "arm64",
@@ -42,7 +46,7 @@ fn target_name(self) -> &'static str {
 
     pub(crate) fn target_arch(self) -> crate::spec::Arch {
         match self {
-            Self::Armv7k | Self::Armv7s => crate::spec::Arch::Arm,
+            Self::Armv6 | Self::Armv7 | Self::Armv7k | Self::Armv7s => crate::spec::Arch::Arm,
             Self::Arm64 | Self::Arm64e | Self::Arm64_32 => crate::spec::Arch::AArch64,
             Self::I386 | Self::I686 => crate::spec::Arch::X86,
             Self::X86_64 | Self::X86_64h => crate::spec::Arch::X86_64,
@@ -51,8 +55,10 @@ pub(crate) fn target_arch(self) -> crate::spec::Arch {
 
     fn target_cpu(self, env: TargetEnv) -> &'static str {
         match self {
+            Self::Armv6 => "mpcore",
+            Self::Armv7 => "cortex-a8",
             Self::Armv7k => "cortex-a8",
-            Self::Armv7s => "swift", // iOS 10 is only supported on iPhone 5 or higher.
+            Self::Armv7s => "swift",
             Self::Arm64 => match env {
                 TargetEnv::Normal => "apple-a7",
                 TargetEnv::Simulator => "apple-a12",
@@ -60,21 +66,18 @@ fn target_cpu(self, env: TargetEnv) -> &'static str {
             },
             Self::Arm64e => "apple-a12",
             Self::Arm64_32 => "apple-s4",
-            // Only macOS 10.12+ is supported, which means
-            // all x86_64/x86 CPUs must be running at least penryn
-            // https://github.com/llvm/llvm-project/blob/01f924d0e37a5deae51df0d77e10a15b63aa0c0f/clang/lib/Driver/ToolChains/Arch/X86.cpp#L79-L82
-            Self::I386 | Self::I686 => "penryn",
-            Self::X86_64 => "penryn",
+            Self::I386 | Self::I686 => "pentium-m",
+            Self::X86_64 => "core2",
             // Note: `core-avx2` is slightly more advanced than `x86_64h`, see
             // comments (and disabled features) in `x86_64h_apple_darwin` for
-            // details. It is a higher baseline then `penryn` however.
+            // details. It is a higher baseline then `core2` however.
             Self::X86_64h => "core-avx2",
         }
     }
 
     fn stack_probes(self) -> StackProbeType {
         match self {
-            Self::Armv7k | Self::Armv7s => StackProbeType::None,
+            Self::Armv6 | Self::Armv7 | Self::Armv7k | Self::Armv7s => StackProbeType::None,
             Self::Arm64
             | Self::Arm64e
             | Self::Arm64_32
@@ -141,22 +144,15 @@ pub(crate) fn base(
         families: cvs!["unix"],
         is_like_darwin: true,
         binary_format: BinaryFormat::MachO,
-        // LLVM notes that macOS 10.11+ and iOS 9+ default
-        // to v4, so we do the same.
-        // https://github.com/llvm/llvm-project/blob/378778a0d10c2f8d5df8ceff81f95b6002984a4b/clang/lib/Driver/ToolChains/Darwin.cpp#L1203
-        default_dwarf_version: 4,
+        default_dwarf_version: 2,
         frame_pointer: match arch {
-            // clang ignores `-fomit-frame-pointer` for Armv7, it only accepts `-momit-leaf-frame-pointer`
-            Arch::Armv7k | Arch::Armv7s => FramePointer::Always,
-            // clang supports omitting frame pointers for the rest, but... don't?
-            Arch::Arm64 | Arch::Arm64e | Arch::Arm64_32 => FramePointer::NonLeaf,
+            Arch::Armv6 | Arch::Armv7 | Arch::Armv7k | Arch::Armv7s | Arch::Arm64 | Arch::Arm64e | Arch::Arm64_32 => FramePointer::NonLeaf,
             Arch::I386 | Arch::I686 | Arch::X86_64 | Arch::X86_64h => FramePointer::Always,
         },
         has_rpath: true,
         dll_suffix: ".dylib".into(),
         archive_format: "darwin".into(),
-        // Thread locals became available with iOS 8 and macOS 10.7,
-        // and both are far below our minimum.
+        // Thread locals became available with iOS 8 and macOS 10.7.
         has_thread_local: true,
         abi_return_struct_as_int: true,
         emit_debug_gdb_scripts: false,
@@ -303,10 +299,10 @@ pub fn os_minimum_deployment_target(os: &Os) -> Self {
         // $ rustc --print deployment-target
         // ```
         let (major, minor, patch) = match os {
-            Os::MacOs => (10, 12, 0),
-            Os::IOs => (10, 0, 0),
-            Os::TvOs => (10, 0, 0),
-            Os::WatchOs => (5, 0, 0),
+            Os::MacOs => (10, 5, 0),
+            Os::IOs => (2, 0, 0),
+            Os::TvOs => (8, 0, 0),
+            Os::WatchOs => (2, 0, 0),
             Os::VisionOs => (1, 0, 0),
             other => {
                 unreachable!("tried to get deployment target for non-Apple platform: {:?}", other)
@@ -325,9 +321,13 @@ pub fn os_minimum_deployment_target(os: &Os) -> Self {
     pub fn minimum_deployment_target(target: &Target) -> Self {
         let (major, minor, patch) = match (&target.os, &target.arch, &target.env) {
             (Os::MacOs, crate::spec::Arch::AArch64, _) => (11, 0, 0),
+            (Os::MacOs, crate::spec::Arch::X86_64, _) => (10, 6, 0),
             (Os::IOs, crate::spec::Arch::AArch64, Env::MacAbi) => (14, 0, 0),
             (Os::IOs, crate::spec::Arch::AArch64, Env::Sim) => (14, 0, 0),
             (Os::IOs, _, _) if target.llvm_target.starts_with("arm64e") => (14, 0, 0),
+            (Os::IOs, _, _) if target.llvm_target.starts_with("arm64") => (7, 0, 0),
+            (Os::IOs, _, _) if target.llvm_target.starts_with("armv7s") => (6, 0, 0),
+            (Os::IOs, _, _) if target.llvm_target.starts_with("armv7") => (3, 0, 0),
             // Mac Catalyst defaults to 13.1 in Clang.
             (Os::IOs, _, Env::MacAbi) => (13, 1, 0),
             (Os::TvOs, crate::spec::Arch::AArch64, Env::Sim) => (14, 0, 0),
diff --git a/compiler/rustc_target/src/spec/mod.rs b/compiler/rustc_target/src/spec/mod.rs
index 424026bdcea..55bd0884030 100644
--- a/compiler/rustc_target/src/spec/mod.rs
+++ b/compiler/rustc_target/src/spec/mod.rs
@@ -1575,6 +1575,8 @@ fn $module() {
     ("aarch64-apple-ios", aarch64_apple_ios),
     ("arm64e-apple-ios", arm64e_apple_ios),
     ("armv7s-apple-ios", armv7s_apple_ios),
+    ("armv7-apple-ios", armv7_apple_ios),
+    ("armv6-apple-ios", armv6_apple_ios),
     ("x86_64-apple-ios-macabi", x86_64_apple_ios_macabi),
     ("aarch64-apple-ios-macabi", aarch64_apple_ios_macabi),
     ("aarch64-apple-ios-sim", aarch64_apple_ios_sim),
diff --git a/compiler/rustc_target/src/spec/targets/aarch64_apple_ios.rs b/compiler/rustc_target/src/spec/targets/aarch64_apple_ios.rs
index 4b8356cc71c..9945494012e 100644
--- a/compiler/rustc_target/src/spec/targets/aarch64_apple_ios.rs
+++ b/compiler/rustc_target/src/spec/targets/aarch64_apple_ios.rs
@@ -1,5 +1,5 @@
 use crate::spec::base::apple::{Arch, TargetEnv, base};
-use crate::spec::{Os, SanitizerSet, Target, TargetMetadata, TargetOptions};
+use crate::spec::{Cc, LinkerFlavor, Lld, Os, SanitizerSet, Target, TargetMetadata, TargetOptions};
 
 pub(crate) fn target() -> Target {
     let (opts, llvm_target, arch) = base(Os::IOs, Arch::Arm64, TargetEnv::Normal);
@@ -21,6 +21,8 @@ pub(crate) fn target() -> Target {
             supported_sanitizers: SanitizerSet::ADDRESS
                 | SanitizerSet::THREAD
                 | SanitizerSet::REALTIME,
+            has_thread_local: false,
+            late_link_args: TargetOptions::link_args(LinkerFlavor::Darwin(Cc::Yes, Lld::No), &["-lsyscompat"]),
             ..opts
         },
     }
diff --git a/compiler/rustc_target/src/spec/targets/armv6_apple_ios.rs b/compiler/rustc_target/src/spec/targets/armv6_apple_ios.rs
new file mode 100644
index 00000000000..5622546e6ff
--- /dev/null
+++ b/compiler/rustc_target/src/spec/targets/armv6_apple_ios.rs
@@ -0,0 +1,25 @@
+use crate::spec::base::apple::{Arch, TargetEnv, base};
+use crate::spec::{Cc, LinkerFlavor, Lld, Os, Target, TargetMetadata, TargetOptions};
+
+pub(crate) fn target() -> Target {
+    let (opts, llvm_target, arch) = base(Os::IOs, Arch::Armv6, TargetEnv::Normal);
+    Target {
+        llvm_target,
+        metadata: TargetMetadata {
+            description: Some("ARMv6K Apple iOS".into()),
+            tier: Some(3),
+            host_tools: Some(false),
+            std: Some(true),
+        },
+        pointer_width: 32,
+        data_layout: "e-m:o-p:32:32-Fi8-f64:32:64-v64:32:64-v128:32:128-a:0:32-n32-S32".into(),
+        arch,
+        options: TargetOptions {
+            features: "+v6,+vfp2".into(),
+            max_atomic_width: Some(64),
+            has_thread_local: false,
+            late_link_args: TargetOptions::link_args(LinkerFlavor::Darwin(Cc::Yes, Lld::No), &["-lgcc_s.1", "-lsyscompat"]),
+            ..opts
+        },
+    }
+}
diff --git a/compiler/rustc_target/src/spec/targets/armv7_apple_ios.rs b/compiler/rustc_target/src/spec/targets/armv7_apple_ios.rs
new file mode 100644
index 00000000000..c20fac6d858
--- /dev/null
+++ b/compiler/rustc_target/src/spec/targets/armv7_apple_ios.rs
@@ -0,0 +1,25 @@
+use crate::spec::base::apple::{Arch, TargetEnv, base};
+use crate::spec::{Cc, LinkerFlavor, Lld, Os, Target, TargetMetadata, TargetOptions};
+
+pub(crate) fn target() -> Target {
+    let (opts, llvm_target, arch) = base(Os::IOs, Arch::Armv7, TargetEnv::Normal);
+    Target {
+        llvm_target,
+        metadata: TargetMetadata {
+            description: Some("ARMv7-A Apple iOS".into()),
+            tier: Some(3),
+            host_tools: Some(false),
+            std: Some(true),
+        },
+        pointer_width: 32,
+        data_layout: "e-m:o-p:32:32-Fi8-f64:32:64-v64:32:64-v128:32:128-a:0:32-n32-S32".into(),
+        arch,
+        options: TargetOptions {
+            features: "+v7,+vfp3,+neon".into(),
+            max_atomic_width: Some(64),
+            has_thread_local: false,
+            late_link_args: TargetOptions::link_args(LinkerFlavor::Darwin(Cc::Yes, Lld::No), &["-lgcc_s.1", "-lsyscompat"]),
+            ..opts
+        },
+    }
+}
diff --git a/compiler/rustc_target/src/spec/targets/armv7s_apple_ios.rs b/compiler/rustc_target/src/spec/targets/armv7s_apple_ios.rs
index cd9da114acf..666abe7fc7d 100644
--- a/compiler/rustc_target/src/spec/targets/armv7s_apple_ios.rs
+++ b/compiler/rustc_target/src/spec/targets/armv7s_apple_ios.rs
@@ -1,5 +1,5 @@
 use crate::spec::base::apple::{Arch, TargetEnv, base};
-use crate::spec::{Os, Target, TargetMetadata, TargetOptions};
+use crate::spec::{Cc, LinkerFlavor, Lld, Os, Target, TargetMetadata, TargetOptions};
 
 pub(crate) fn target() -> Target {
     let (opts, llvm_target, arch) = base(Os::IOs, Arch::Armv7s, TargetEnv::Normal);
@@ -17,6 +17,8 @@ pub(crate) fn target() -> Target {
         options: TargetOptions {
             features: "+v7,+vfp4,+neon".into(),
             max_atomic_width: Some(64),
+            has_thread_local: false,
+            late_link_args: TargetOptions::link_args(LinkerFlavor::Darwin(Cc::Yes, Lld::No), &["-lsyscompat"]),
             ..opts
         },
     }
diff --git a/compiler/rustc_target/src/spec/targets/i686_apple_darwin.rs b/compiler/rustc_target/src/spec/targets/i686_apple_darwin.rs
index 1c3481928a8..4f9ae177a7a 100644
--- a/compiler/rustc_target/src/spec/targets/i686_apple_darwin.rs
+++ b/compiler/rustc_target/src/spec/targets/i686_apple_darwin.rs
@@ -1,12 +1,12 @@
 use crate::spec::base::apple::{Arch, TargetEnv, base};
-use crate::spec::{Os, Target, TargetMetadata, TargetOptions};
+use crate::spec::{Cc, LinkerFlavor, Lld, Os, Target, TargetMetadata, TargetOptions};
 
 pub(crate) fn target() -> Target {
     let (opts, llvm_target, arch) = base(Os::MacOs, Arch::I686, TargetEnv::Normal);
     Target {
         llvm_target,
         metadata: TargetMetadata {
-            description: Some("x86 Apple macOS (10.12+, Sierra+)".into()),
+            description: Some("x86 Apple macOS (10.5+, Leopard+)".into()),
             tier: Some(3),
             host_tools: Some(true),
             std: Some(true),
@@ -16,6 +16,12 @@ pub(crate) fn target() -> Target {
             i128:128-f64:32:64-f80:128-n8:16:32-S128"
             .into(),
         arch,
-        options: TargetOptions { mcount: "\u{1}mcount".into(), max_atomic_width: Some(64), ..opts },
+        options: TargetOptions {
+            mcount: "\u{1}mcount".into(),
+            max_atomic_width: Some(64),
+            has_thread_local: false,
+            late_link_args: TargetOptions::link_args(LinkerFlavor::Darwin(Cc::Yes, Lld::No), &["-lgcc_s.1", "-lsyscompat"]),
+            ..opts
+        },
     }
 }
diff --git a/compiler/rustc_target/src/spec/targets/x86_64_apple_darwin.rs b/compiler/rustc_target/src/spec/targets/x86_64_apple_darwin.rs
index 8d0febc362d..3ce44ab73a3 100644
--- a/compiler/rustc_target/src/spec/targets/x86_64_apple_darwin.rs
+++ b/compiler/rustc_target/src/spec/targets/x86_64_apple_darwin.rs
@@ -1,12 +1,12 @@
 use crate::spec::base::apple::{Arch, TargetEnv, base};
-use crate::spec::{Os, SanitizerSet, Target, TargetMetadata, TargetOptions};
+use crate::spec::{Cc, LinkerFlavor, Lld, Os, SanitizerSet, Target, TargetMetadata, TargetOptions};
 
 pub(crate) fn target() -> Target {
     let (opts, llvm_target, arch) = base(Os::MacOs, Arch::X86_64, TargetEnv::Normal);
     Target {
         llvm_target,
         metadata: TargetMetadata {
-            description: Some("x86_64 Apple macOS (10.12+, Sierra+)".into()),
+            description: Some("x86_64 Apple macOS (10.6+, Snow Leopard+)".into()),
             tier: Some(2),
             host_tools: Some(true),
             std: Some(true),
@@ -17,13 +17,15 @@ pub(crate) fn target() -> Target {
         arch,
         options: TargetOptions {
             mcount: "\u{1}mcount".into(),
-            max_atomic_width: Some(128), // penryn+ supports cmpxchg16b
+            max_atomic_width: Some(128), // core2+ supports cmpxchg16b
             supported_sanitizers: SanitizerSet::ADDRESS
                 | SanitizerSet::CFI
                 | SanitizerSet::LEAK
                 | SanitizerSet::THREAD
                 | SanitizerSet::REALTIME,
             supports_xray: true,
+            has_thread_local: false,
+            late_link_args: TargetOptions::link_args(LinkerFlavor::Darwin(Cc::Yes, Lld::No), &["-lsyscompat"]),
             ..opts
         },
     }
diff --git a/compiler/rustc_target/src/spec/targets/x86_64h_apple_darwin.rs b/compiler/rustc_target/src/spec/targets/x86_64h_apple_darwin.rs
index 7f1fe485aa1..c71898262a0 100644
--- a/compiler/rustc_target/src/spec/targets/x86_64h_apple_darwin.rs
+++ b/compiler/rustc_target/src/spec/targets/x86_64h_apple_darwin.rs
@@ -1,5 +1,5 @@
 use crate::spec::base::apple::{Arch, TargetEnv, base};
-use crate::spec::{Os, SanitizerSet, Target, TargetMetadata, TargetOptions};
+use crate::spec::{Cc, LinkerFlavor, Lld, Os, SanitizerSet, Target, TargetMetadata, TargetOptions};
 
 pub(crate) fn target() -> Target {
     let (mut opts, llvm_target, arch) = base(Os::MacOs, Arch::X86_64h, TargetEnv::Normal);
@@ -37,6 +37,10 @@ pub(crate) fn target() -> Target {
         data_layout:
             "e-m:o-p270:32:32-p271:32:32-p272:64:64-i64:64-i128:128-f80:128-n8:16:32:64-S128".into(),
         arch,
-        options: TargetOptions { mcount: "\u{1}mcount".into(), ..opts },
+        options: TargetOptions {
+            mcount: "\u{1}mcount".into(),
+            late_link_args: TargetOptions::link_args(LinkerFlavor::Darwin(Cc::Yes, Lld::No), &["-lsyscompat"]),
+            ..opts
+        },
     }
 }
diff --git a/library/std/Cargo.toml b/library/std/Cargo.toml
index 3a673cf23b8..8466f9e572c 100644
--- a/library/std/Cargo.toml
+++ b/library/std/Cargo.toml
@@ -10,7 +10,7 @@ edition = "2024"
 autobenches = false
 
 [lib]
-crate-type = ["dylib", "rlib"]
+crate-type = ["rlib"]
 
 [dependencies]
 alloc = { path = "../alloc", public = true }
diff --git a/library/std/src/sys/process/unix/unix.rs b/library/std/src/sys/process/unix/unix.rs
index 7d944f2f7ee..cb6adb756b8 100644
--- a/library/std/src/sys/process/unix/unix.rs
+++ b/library/std/src/sys/process/unix/unix.rs
@@ -5,8 +5,6 @@
 #[cfg(not(any(
     target_os = "vxworks",
     target_os = "l4re",
-    target_os = "tvos",
-    target_os = "watchos",
 )))]
 use libc::{gid_t, uid_t};
 
@@ -164,26 +162,9 @@ pub fn spawn(
         }
     }
 
-    // WatchOS and TVOS headers mark the `fork`/`exec*` functions with
-    // `__WATCHOS_PROHIBITED __TVOS_PROHIBITED`, and indicate that the
-    // `posix_spawn*` functions should be used instead. It isn't entirely clear
-    // what `PROHIBITED` means here (e.g. if calls to these functions are
-    // allowed to exist in dead code), but it sounds bad, so we go out of our
-    // way to avoid that all-together.
-    #[cfg(any(target_os = "tvos", target_os = "watchos"))]
-    const ERR_APPLE_TV_WATCH_NO_FORK_EXEC: Error = io::const_error!(
-        ErrorKind::Unsupported,
-        "`fork`+`exec`-based process spawning is not supported on this target",
-    );
-
-    #[cfg(any(target_os = "tvos", target_os = "watchos"))]
-    unsafe fn do_fork(&mut self) -> Result<pid_t, io::Error> {
-        return Err(Self::ERR_APPLE_TV_WATCH_NO_FORK_EXEC);
-    }
-
     // Attempts to fork the process. If successful, returns Ok((0, -1))
     // in the child, and Ok((child_pid, -1)) in the parent.
-    #[cfg(not(any(target_os = "watchos", target_os = "tvos", target_os = "nto")))]
+    #[cfg(not(target_os = "nto"))]
     unsafe fn do_fork(&mut self) -> Result<pid_t, io::Error> {
         cvt(libc::fork())
     }
@@ -274,7 +255,6 @@ pub fn exec(&mut self, default: Stdio) -> io::Error {
     // allocation). Instead we just close it manually. This will never
     // have the drop glue anyway because this code never returns (the
     // child will either exec() or invoke libc::exit)
-    #[cfg(not(any(target_os = "tvos", target_os = "watchos")))]
     unsafe fn do_exec(
         &mut self,
         stdio: ChildPipes,
@@ -409,15 +389,6 @@ fn drop(&mut self) {
         Err(io::Error::last_os_error())
     }
 
-    #[cfg(any(target_os = "tvos", target_os = "watchos"))]
-    unsafe fn do_exec(
-        &mut self,
-        _stdio: ChildPipes,
-        _maybe_envp: Option<&CStringArray>,
-    ) -> Result<!, io::Error> {
-        return Err(Self::ERR_APPLE_TV_WATCH_NO_FORK_EXEC);
-    }
-
     #[cfg(not(any(
         target_os = "freebsd",
         target_os = "illumos",
diff --git a/library/std/src/sys/sync/thread_parking/mod.rs b/library/std/src/sys/sync/thread_parking/mod.rs
index 74b5b72b19a..ef622621971 100644
--- a/library/std/src/sys/sync/thread_parking/mod.rs
+++ b/library/std/src/sys/sync/thread_parking/mod.rs
@@ -26,11 +26,6 @@
         mod windows7;
         pub use windows7::Parker;
     }
-    all(target_vendor = "apple", not(miri)) => {
-        // Doesn't work in Miri, see <https://github.com/rust-lang/miri/issues/2589>.
-        mod darwin;
-        pub use darwin::Parker;
-    }
     target_os = "xous" => {
         mod xous;
         pub use xous::Parker;
